docs/fundamentals/gateway/custom-gateway.md

```diff
@@ -0,0 +1,324 @@
+        await self.server.serve()
+```
+* Implement `async def shutdown():`. This should stop the server and free all resources associated with it:
+```python
+from jina import Gateway
+
+
+class MyGateway(Gateway):
+    ...
+
+    async def shutdown(self):
+        self.server.should_exit = True
+        await self.server.shutdown()
+```
+
+As an example, you can refer to {class}`~jina.serve.runtimes.gateway.grpc.GRPCGateway` and 
+{class}`~jina.serve.runtimes.gateway.websocket.WebSocketGateway`.
+
+### Using {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`:
+{class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway` offers a simpler API to implement custom 
+gateways but is restricted to FastAPI apps.
+
+In order to implement a custom gateway using {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`, 
+simply implement the {meth}`~jina.jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway.app` property:
+
+```python
+from jina.serve.runtimes.gateway.http.fastapi import FastAPIBaseGateway
+
+
+class MyGateway(FastAPIBaseGateway):
+    @property
+    def app(self):
+        from fastapi import FastAPI
+
+        app = FastAPI(title='Custom FastAPI Gateway')
+
+        @app.get(path='/endpoint')
+        def custom_endpoint():
+            return {'message': 'custom-fastapi-gateway'}
+
+        return app
+```
+
+As an example, you can refer to {class}`~jina.serve.runtimes.gateway.http.HTTPGateway`.
+
+{class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway` is a subclass of {class}`~jina.Gateway` and therefore 
+shares its attributes.
+
+## Gateway arguments
+### Runtime attributes
+Jina injects some attributes into the Gateway classes. You can use them to set up your custom gateway:
+* logger: Jina logger object.
+* streamer: {class}`~jina.serve.streamer.GatewayStreamer`. Use this object to send Documents from the Gateway to Executors. Refer to {ref}`this section <gateway-streamer>` for more information.
+* runtime_args: `argparse.Namespace` object containing runtime arguments.
+* port: main port exposed by the Gateway.
+* ports: list all ports exposed by the Gateway.
+* protocols: list all protocols supported by the Gateway.
+* host: host address to which the Gateway server should be bound.
+
+```{admonition} Nonte
+:class: note
+
+Jina provides the Gateway with a list of ports and protocols to expose. Therefore, a custom Gateway can handle requests 
+on multiple ports using different protocols.
+```
+
+### User-defined parameters
+You can also set other parameters by implementing a custom constructor `__init__`.You can also override constructor 
+parameters in the Flow Python API (using `uses_with` parameter) or in the YAML configuration when including the gateway.
+Refer to the {ref}`Use Custom Gateway section <use-custom-gateway>` for more information.
+
+(gateway-streamer)=
+## Calling Executors with {class}`~jina.serve.streamer.GatewayStreamer`
```

Comment:
I would move this section further up, to me it is the key thing that the user needs to worry about and work with. In comparison, the other injected attributes mentioned above seem like a detail / less important.